R/utils.R

```diff
@@ -140,3 +140,101 @@ edit.data.table = function(name, ...) {
+      if (length(colsub) == 2L && colsub[[1L]] == "-" && is.numeric(colsub[[2L]])) {
+        negate_cols = TRUE ##what about -(-1)???
+        colsub = colsub[[2L]]
+      }
+    }
+    if (length(colsub) == 3L && colsub[[1L]] == ":") {
+      if (is.name(colsub[[2L]]) && is.name(colsub[[3L]])){
+        # cols is of the format a:c
+        rnge = chmatch(c(as.character(colsub[[2L]]), as.character(colsub[[3L]])), names(x))
+        if (anyNA(rnge)) stop(gettextf("Not all columns were found in the %s column select statement: [%s]", mode, deparse(origsub), domain = "R-data.table"))
+        cols = rnge[1L]:rnge[2L] 
+      } else if (is.character(colsub[[2L]]) || is.character(colsub[[3L]])){
+        stop(gettextf("When selecting a range of columns in the var_1:var_2 format, quotation marks should not be used. Instead of %s try %s:%s", deparse(origsub), deparse(as.name(colsub[[2L]])), deparse(as.name(colsub[[3L]])), domain = "R-data.table"))
+      } else {
+        # cols is of the format 1:3 or would also allow min(var):max(var)
+        cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+      }
+    } else if (length(colsub) > 1 && colsub[[1L]] == "patterns") {
+      # each pattern gives a new filter condition, intersect the end result
+      x_names = names(x)
+      cols = Reduce(intersect, lapply(as.list(colsub)[-1L], function(col) grep(eval(col, parent.frame(2L)), x_names)))
+    } else {
+      cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+    }
+  } else {
+    # if colsub is also a name in the list, we prioritize it over anything in the environment
+    if (!exists(cols <- as.character(colsub), where = x)) 
+      cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+  }
+  
+  # allow filtering via function in .SDcols, #3950
+  if (is.function(cols)) {
+    cols = lapply(x, cols)
+    if (any(idx <- vapply_1i(cols, length) > 1L | vapply_1c(cols, typeof) != 'logical' | vapply_1b(cols, anyNA)))
+      stop(gettextf("When %s is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met for: %s", mode, brackify(names(x)[idx]), domain = "R-data.table"))
+    cols = unlist(cols, use.names = FALSE)
+  }
+  
+  if (anyNA(cols))
+    stop(gettextf("%s missing at the following indices: %s", mode, brackify(which(is.na(cols)))))
+  
+  if (is.logical(cols)) {
+    if ((col_len <- length(cols)) == 1L) {
+      cols = rep(cols, length.out = x_len)
+    } else if (col_len != x_len) {
+      ## TODO change to error in 2022
+      warning(gettextf("When %s is a logical vector, each column should have a TRUE or FALSE entry. The current logical vector of length %d will be repeated to length of data.table. Warning will change to error in the next verion.", mode, col_len, domain = "R-data.table"))
+      cols = rep(cols, length.out = x_len)
+    } 
```

Comment:
RE: recycling logical vectors. Base recycles vectors to the length of the data.frame. See ```iris[, TRUE]``` or ```iris[, c(TRUE, FALSE)]```. I guess I'm leaning towards allowing vectors repeat.

But more broadly, I am interested in some way to select all the variables. ```colsub = TRUE``` seemed like a quick and easy way. ```names(.SD)``` could work but if this function was ever applied to ```duplicated``` or ```melt``` or any other functions that allow column selection, I am not sure ```names(.SD)``` would make sense. Maybe a made up call ```all_cols()``` that we could evaluate?

Reply:
Your use of colsub=TRUE can be achieved by not passing SDcols, or passing names, or rep(TRUE, ncol(DT)) or seq_along. Most of which requires you to refer to DT, which is not chaining friendly. You can always use a function (...) TRUE to workaround this limitation.
In the issue related to logical vector in SDcols we discussed recycling and so far conclusion whas to not recycle.